lib/service/validateconfig.go

```diff
@@ -0,0 +1,148 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"io"
+	"strings"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/trace"
+
+	"github.com/sirupsen/logrus"
+)
+
+func validateConfig(cfg *Config) error {
+	applyDefaults(cfg)
+
+	if err := validateVersion(cfg); err != nil {
+		return err
+	}
+
+	if err := verifyEnabledService(cfg); err != nil {
+		return err
+	}
+
+	if err := validateAuthOrProxyServices(cfg); err != nil {
+		return err
+	}
+
+	if cfg.DataDir == "" {
+		return trace.BadParameter("config: please supply data directory")
+	}
+
+	for i := range cfg.Auth.Authorities {
+		if err := services.ValidateCertAuthority(cfg.Auth.Authorities[i]); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	for _, tun := range cfg.ReverseTunnels {
+		if err := services.ValidateReverseTunnel(tun); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	cfg.SSH.Namespace = types.ProcessNamespace(cfg.SSH.Namespace)
+
+	return nil
+}
+
+func applyDefaults(cfg *Config) {
+	if cfg.Version == "" {
+		cfg.Version = defaults.TeleportConfigVersionV1
+	}
+
+	if cfg.Console == nil {
+		cfg.Console = io.Discard
+	}
+
+	if cfg.Log == nil {
+		cfg.Log = logrus.StandardLogger()
+	}
+
+	if cfg.PollingPeriod == 0 {
+		cfg.PollingPeriod = defaults.LowResPollingPeriod
+	}
+}
+
+func validateAuthOrProxyServices(cfg *Config) error {
+	haveAuthServers := len(cfg.authServers) > 0
+	haveProxyAddress := !cfg.ProxyAddress.IsEmpty()
+
+	if cfg.Version == defaults.TeleportConfigVersionV3 {
+		if haveAuthServers && haveProxyAddress {
+			return trace.BadParameter("config: cannot use both auth_server and proxy_address")
+		}
+
+		if !haveAuthServers && !haveProxyAddress {
+			return trace.BadParameter("config: auth_server or proxy_address is required")
+		}
+
+		return nil
+	}
+
+	if haveProxyAddress {
+		return trace.BadParameter("config: proxy_address is supported from config version v3 onwards")
+	}
+
+	if !haveAuthServers {
+		return trace.BadParameter("config: auth_servers is required")
+	}
+
+	return nil
+}
+
+func validateVersion(cfg *Config) error {
+	hasVersion := utils.SliceContainsStr(defaults.TeleportVersions, cfg.Version)
+	if !hasVersion {
+		return trace.BadParameter("config: version must be one of %s", strings.Join(defaults.TeleportVersions, ", "))
+	}
+
+	return nil
+}
+
+func verifyEnabledService(cfg *Config) error {
+	enabled := []bool{
+		cfg.Auth.Enabled,
+		cfg.SSH.Enabled,
+		cfg.Proxy.Enabled,
+		cfg.Kube.Enabled,
+		cfg.Apps.Enabled,
+		cfg.Databases.Enabled,
+		cfg.WindowsDesktop.Enabled,
```

Comment:
I think we just added another "discovery" service that you'll probably want to add here too.